sys/include/net/gcoap/dns.h

```diff
@@ -0,0 +1,200 @@
+ * @return  -ENOBUFS, if length of received CoAP body is greater than
+ *          @ref CONFIG_DNS_MSG_LEN.
+ * @return  -ENOENT, if Zone-ID of the URI can not be found locally.
+ * @return  -ENOMSG, if CoAP response did not contain a DNS response.
+ * @return  -ENOTRECOVERABLE, on gCoAP-internal error.
+ * @return  -ENOTSUP, if credential can not be added for to client.
+ * @return  -ENOTSUP, if module `gcoap_dns_async` is not compiled in.
+ * @return  -ETIMEDOUT, if CoAP request timed out.
+ */
+int gcoap_dns_query(const char *domain_name, void *addr_out, int family);
+
+/**
+ * @brief   Sets and checks a URI
+ *
+ * @param[in] uri   A URI. May be NULL to remove the URI.
+ *
+ * @return  length of @p uri on success (0 when URI is removed).
+ * @return  -EINVAL, if @p uri is not a valid URI for DNS over CoAP.
+ * @return  -ENOBUFS, if the client implementation has not enough buffer space
+ *          to process the URI.
+ */
+int gcoap_dns_server_uri_set(const char *uri);
+
+/**
+ * @brief   Gets the URI
+ *
+ * This function is not thread-safe.
+ *
+ * @return  The current URI for the client.
+ */
+const char *gcoap_dns_server_uri_get(void);
+
+/**
+ * @brief   Deletes all added credentials
+ *
+ * This also removes the credentials from the @ref net_credman and the GCoAP DTLS sock.
+ */
+void gcoap_dns_cred_reset(void);
+
+/**
+ * @brief   Adds a credential for the use with the configured DoC server.
+ *
+ * @pre creds != NULL.
+ *
+ * @param[in] creds     A crential. May not be NULL.
+ *
+ * @return 0 on success
+ * @return -ENOTSUP if DTLS is not supported (because `gcoap_dtls` is not compiled in)
+ * @return -ENOMEM if @ref CONFIG_GCOAP_DNS_CREDS_MAX is exceeded (duplicates may be added).
+ * @return -EBADF if the credential can not be added to @ref net_credman or the GCoAP DTLS sock.
+ */
+int gcoap_dns_cred_add(credman_credential_t *creds);
+
+/**
+ * @brief   Remove a credential for the use with the configured DoC server
+ *
+ * This also removes the credentials from the @ref net_credman and the GCoAP DTLS sock.
+ *
+ * @param[in] tag   The tag of the credential.
+ * @param[in] type  The type of the credential.
+ */
+void gcoap_dns_cred_remove(credman_tag_t tag, credman_type_t type);
+
+/**
+ * @brief   Deletes the proxy URI.
+ */
+void gcoap_dns_server_proxy_reset(void);
+
+/**
+ * @brief   Sets and checks a proxy URI.
+ *
+ * @param[in] proxy     A proxy URI. Must contain a `dns` query variable.
+ *
+ * @return  length of @p proxy on success.
+ * @return  -ENOBUFS, if the client implementation has not enough buffer space
+ *          to process the proxy URI.
+ * @return  -ENOSUP, if module `gcoap_dns_proxied` is not compiled in.
+ */
+int gcoap_dns_server_proxy_set(const char *proxy);
+
+/**
+ * @brief   Gets the proxy URI
+ *
+ * This function is not thread-safe.
+ *
+ * @return  The current proxy URI for the client.
+ */
+const char *gcoap_dns_server_proxy_get(void);
```

Comment:
As with `server_uri_get` -- do we need that public?

Reply:
Same proposal as in https://github.com/RIOT-OS/RIOT/pull/16705/files#r908797456 would then also apply here ;-).